### PR TITLE
Xml

### PR DESCRIPTION
**Pull request by cwulfman**
pulling the cvrs and edf together

### CODE REVIEW

**Comment by cwulfman** [comment]
addresses issue #4

File: data/cvr/samples/jetsons/jetsons_main_edf.xml [hunk 1]
```diff
@@ -12,7 +16,12 @@
                 </OrderedContent>
             </OrderedContent>
         </BallotStyle>
+
         <BallotStyle>
+            <ExternalIdentifier>
+                <Type>local-level</Type>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_main_edf.xml:22]: I hadn't thought of using `local-level` here. I'm pretty sure that makes sense. I'll change the JSON accordingly here and for the other `ExternalIdentifer`s.

File: data/cvr/samples/jetsons/jetsons_main_edf.xml [hunk 2]
```diff
@@ -99,30 +117,98 @@
...
+                <Value>gc-mayor-orbit-city</Value>
+            </ExternalIdentifier>
             <Name>Mayor of Orbit City</Name>
             <VoteVariation>plurality</VoteVariation>
             <VotesAllowed>1</VotesAllowed>
         </Contest>
+
         <Contest xsi:type="CandidateContest" ObjectId="contest-control-board-spaceport">
+            <ContestSelection xsi:type="CandidateSelection"
+                ObjectId="contest-control-board-spaceport-selection-jane-jetson">
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_main_edf.xml]: See note "ContestSelectionId Separators" at the top of the review.

File: data/cvr/samples/jetsons/jetsons_main_edf.xml [hunk 3]
```diff
@@ -99,30 +117,98 @@
...
+            <ContestSelection xsi:type="CandidateSelection"
+                ObjectId="contest-mayor-selection-write-in">
+                <SequenceOrder>3</SequenceOrder>
+                <IsWriteIn>true</IsWriteIn>
+            </ContestSelection>
+
             <ElectionDistrictId>gadget-county</ElectionDistrictId>
+            <ExternalIdentifier>
+                <Type>local-level</Type>
+                <Value>gc-mayor-orbit-city</Value>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_main_edf.xml]: Use *`gadget-county`* `-mayor-orbit-city`

File: data/cvr/samples/jetsons/jetsons_main_edf.xml [hunk 4]
```diff
@@ -99,30 +117,98 @@
...
+            </ContestSelection>
+
+            <ContestSelection xsi:type="CandidateSelection"
+                ObjectId="contest-mayor-selection-write-in">
+                <SequenceOrder>3</SequenceOrder>
+                <IsWriteIn>true</IsWriteIn>
+            </ContestSelection>
+
             <ElectionDistrictId>gadget-county</ElectionDistrictId>
+            <ExternalIdentifier>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_main_edf.xml:148]: Missing from the JSON side.

File: data/cvr/samples/jetsons/jetsons_main_edf.xml [hunk 5]
```diff
@@ -99,30 +117,98 @@
...
+
             <ElectionDistrictId>gadget-county</ElectionDistrictId>
             <Name>Spaceport Control Board</Name>
             <VoteVariation>n-of-m</VoteVariation>
             <VotesAllowed>2</VotesAllowed>
         </Contest>
+
         <Contest xsi:type="BallotMeasureContest" ObjectId="contest-ballot-measure-gadget-county-1">
             <ElectionDistrictId>gadget-county</ElectionDistrictId>
+            <ExternalIdentifier>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_main_edf.xml:200]: Missing from the JSON side.

File: data/cvr/samples/jetsons/jetsons_main_edf.xml [hunk 6]
```diff
@@ -99,30 +117,98 @@
...
+                <Type>local-level</Type>
+                <Value>contest-gadget-county-ballot-measure-1</Value>
+            </ExternalIdentifier>
             <Name>Shall Gadget County increase its sales tax from 1% to 1.1% for the purpose of raising additional revenue to fund expanded air traffic control operations?</Name>
         </Contest>
 
         <ElectionScopeId>gadget-county</ElectionScopeId>
+        <ExternalIdentifier>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_main_edf.xml:208]: Missing from the JSON side.

File: data/cvr/samples/jetsons/jetsons_main_edf.xml [hunk 7]
```diff
@@ -99,30 +117,98 @@
...
+            <Type>local-level</Type>
+            <Value>gadget_county_election</Value>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_main_edf.xml]: Use `election-gadget-county`.

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 8]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<CastVoteRecordReport xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+ xsi:schemaLocation="http://itl.nist.gov/ns/voting/1500-103/v1 ../../../../../CastVoteRecords/NIST_V0_cast_vote_records.xsd"
+ xmlns="http://itl.nist.gov/ns/voting/1500-103/v1">
+    <CVR>
+        <BallotImage>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: See note on "Ballot Images" at the top of the review.

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            <Image FileName="bedrock_ballot_1" MimeType="application/pdf" />
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: Use: `ballot-image-bedrock-precinct-01.pdf`?
This a filename and not an ID, so arguably the convention can be different. But I'd like to make them visually distinct so if not that then `Bedrock_Ballot_01.pdf` would be better.

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        </BallotImage>
+        <BallotStyleId>bedrock-precinct-ballot</BallotStyleId>
+        <CurrentSnapshotId>sn_0</CurrentSnapshotId>
+        
+
+        <!-- Base snapshot is derived from the output from the mobile app -->
+        <CVRSnapshot ObjectId="sn_0">
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: Use: `snapshot-01`. It's probably not likely there will be more than 9 snapshots but it's not impossible either.

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            <CVRContest>
+                <ContestId>ballot_measure_1</ContestId>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: Use: `contest-ballot-measure-gadget-county-1`

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+                <CVRContestSelection>
+                    <ContestSelectionId>contest_selection_yes</ContestSelectionId>
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml:15]: Use: `contest-ballot-measure-1/selection-yes` or `contest-ballot-measure-1--selection-yes`

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            <Type>original</Type>
+        </CVRSnapshot>
+        <ElectionId>gadget-county-2021-06</ElectionId>
+    </CVR>
+    <Election ObjectId="gadget-county-2021-06">
+        <Code>
+            <Type>local-level</Type>
+            <Value>bedrock-precinct</Value>
+        </Code>
+        <Contest ObjectId="ballot_measure_1" xsi:type="BallotMeasureContest">
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: Use: `contest-ballot-measure-gadget-county-1`

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            <Code>
+                <Type>local-level</Type>
+                <Value>contest-gadget-county-ballot-measure-1</Value>
+            </Code>
+            <ContestSelection xsi:type="BallotMeasureSelection" ObjectId="contest_selection_yes">
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: Use: `contest-ballot-measure-1/selection-yes` or `contest-ballot-measure-1--selection-yes`

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+                <Code>
+                    <Type>local-level</Type>
+                    <Value>yes</Value>
+                </Code>
+                <Selection>yes</Selection>
+            </ContestSelection>
+            <ContestSelection xsi:type="BallotMeasureSelection" ObjectId="contest_selection_no">
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: Use: `contest-ballot-measure-1/selection-no` or `contest-ballot-measure-1--selection-no`

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        
+        <ElectionScopeId>bedrock-precinct</ElectionScopeId>
+    </Election>
+    <GeneratedDate>2021-06-01T12:00:00-08:00</GeneratedDate>
+    <GpUnit ObjectId="bedrock-precinct">
+        <Name>bedrock precinct</Name>
+        <Type>precinct</Type>
+    </GpUnit>
+    <ReportGeneratingDeviceIds>_RD1</ReportGeneratingDeviceIds>
+    <ReportingDevice ObjectId="_RD1">
```
Comment (ion-oset) [inline data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: Use: `ballot-marker-1`.

**Review by ion-oset — CHANGES_REQUESTED** [review]
A couple of possible issues that are easier to discuss here, though I did cite specific instances with comments.

### Identifiers

There are a couple of areas where either the XML or the JSON should change their naming convention. I want to use the same convention for all IDs for consistency, both to avoid confusion but especially because at some point we want to generate them.

#### Naming convention

My convention has been to prefix the ID with its type, then the entity it names, and if there's more than one of a kind (e.g. snapshots) to use 2-digit numbers not 1. This won't matter for this demo but may matter for more serious ones. We could change this but we should keep the IDs human readable (that's to make it easier to debug them) and use the same format consistently.

#### Default Separator

The XML uses `'_'` as a separator and the JSON uses `'-'`. For consistency we should use the same separator, because keeping them synced up will get to be nuisance otherwise. If the XML schema allows dashes in IDs then I'd prefer to use them as they are easier to type. Otherwise I'll change the JSON schema to use underscores.

#### ContestSelectionId Separators

The `ContestSelectionId`s use a `/` between the contest and the selection rather than a `-`. I found this made them slightly less likely to be confused with the `ContestId`s. But if I'm understanding correctly XMLSchema defines these ids as [IDREF](https://www.w3.org/TR/2004/REC-xmlschema-2-20041028/datatypes.html#IDREF)s which are XML [NCName](https://www.w3.org/TR/1999/REC-xml-names-19990114/#NT-NCName)s and don't allow slashes. If that's the case I see a handful ways to unify, that might preserve the separation of the two:

   (1) Replace my use of slashes with underscores (`_`).
   (2) Use double dashes (`--`). 
   (3) Just use single dashes (`-`)

   I currently think I favor 2. 1 isn't visually distinctive enough and isn't worth it. 3 is acceptable though I really did find it easier to avoid mistakes with the separation.

### External Identifiers

There are several places where you've added `ExternalIdentifier`s that are missing from the *JSON* side. I added comments to them. They don't need any changes from you, it's just notes to me.

### Ballot images

I seem to recall John telling me we didn't need `BallotImage` elements in the Jetsons CVR. I'm guessing with the sample ballots that's changed?

### Schema File Name

I'd prefer if the XML schema and JSON schema file use the same pattern, though I guess it's not as important as it is for the identifiers. Any objection (stylistic or technical) to using `nist-cvr-v1_xmlschema.xsd` instead? There's no standard JSON Schema suffix or we could use that instead.

File: data/cvr/samples/jetsons/jetsons_main_edf.xml (hunk 2 above)
Comment (cwulfman) [reply data/cvr/samples/jetsons/jetsons_main_edf.xml]: I've changed all of these to contest-X--selction-Y.

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml (hunk 16 above)
Comment (cwulfman) [reply data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: done

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml (hunk 8 above)
Comment (cwulfman) [reply data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: removed all the <BallotImage/> elements

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml (hunk 10 above)
Comment (cwulfman) [reply data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml]: done

File: data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml (hunk 12 above)
Comment (cwulfman) [reply data/cvr/samples/jetsons/jetsons_bedrock-precinct_cvr.xml:15]: done

**Pull request merged**